Fix error with Jetpack Markdown being enabled

mkaz opened the pull request:
## Description

When saving text blocks while Jetpack Markdown is enabled, the `<p>` tags are stripped on save. When loaded again in Gutenberg the blocks are not recognized since the tags are needed to properly parse paragraph blocks.

## How to Test

1. Install Jetpack plugin in Settings, under Compose enable Markdown syntax
2. Create a new post with at least one text block and save
3. Click over to view the post and then click "Edit in Gutenberg"
4. The text block will have paragraph tags deleted and be asked to convert to classic block.

## Change

* This change disables markdown on content that is detected as Gutenberg by adding a filter using the `rest_pre_insert_post` hook.

mkaz committed:
Switch check for ! isset() to just check if array, current logic would not flip settings if already set to true

mkaz committed:
Add function on init to remove jetpack markdown filter

This ends up being a bit heavy handed solution, since it will also
disable markdown in the classic editor, however each block doesn't
fail to classic blocks on each save.

Is there a check we can add to init to see which editor is being used?
The markdown processing is called during save a post, so not sure it is
possible to determine or not

aduth commented on gutenberg.php:
```diff
@@ -3,7 +3,7 @@
  * Plugin Name: Gutenberg
  * Plugin URI: https://github.com/WordPress/gutenberg
  * Description: Printing since 1440. This is the development plugin for the new block editor in core. <strong>Meant for development, do not run on real sites.</strong>
- * Version: 1.2.0
```
The version bumps here shouldn't be part of the changes.

mkaz replied:
Rebased and removed.

aduth commented on lib/compat.php:
```diff
@@ -131,10 +131,22 @@ function gutenberg_ensure_wp_api_request() {
  */
 function gutenberg_disable_editor_settings_wpautop( $settings ) {
-	if ( ! isset( $settings['wpautop'] ) ) {
+	if ( is_array( $settings ) ) {
```
I don't think that `$settings` would ever not be an array. What we might want to do here is test first whether the post contains blocks and, if it does, assign `$settings['wpautop']` as explicitly `false` but otherwise do nothing with the key.

mkaz replied:
I like that, a bit cleaner.

aduth commented on lib/compat.php:
```diff
@@ -131,10 +131,22 @@ function gutenberg_ensure_wp_api_request() {
+ * but disabling does fix the issue
+ */
+function gutenberg_remove_jetpack_markdown_support() {
+	if ( defined( 'JETPACK__VERSION' ) ) {
```
Would it hurt if we removed `wpcom-markdown` support without explicitly testing whether we're in the context of Jetpack, since I assume the post support (by its prefix) is relevant for Jetpack only anyways?

mkaz replied:
I changed the method of determining and removed it regardless of JETPACK being defined.

aduth commented on lib/compat.php:
```diff
@@ -131,10 +131,22 @@ function gutenberg_ensure_wp_api_request() {
+ */
+function gutenberg_remove_jetpack_markdown_support() {
+	if ( defined( 'JETPACK__VERSION' ) ) {
+		remove_post_type_support( 'post', 'wpcom-markdown' );
```
`post` is not the only post type we'd expect to be using Gutenberg. Or is it the case that Jetpack is only adding support for this type? Otherwise we might have to hook into where Jetpack is testing the post type and check whether it's handled by Gutenberg. Currently we have a filter for this, but it's not very well generalized:

https://github.com/WordPress/gutenberg/blob/5a949b705ebc0ec3d05b9c98a653c6cd1697a820/lib/register.php#L154

I could imagine a function to call which runs the filter.

We also want to make sure that we're not removing this support if the post is being edited from the classic editor, since this action will be triggered regardless of whether the user is using the Gutenberg editor.

mkaz committed:
Limit scope of disabling markdown support to just content

Use the pre insert filter on the rest endpoint to determine
if the content being saved is from Gutenberg and disable the
markdown support if a block is included.

Temporary includes the content_has_block() function from PR#2806

mkaz committed:
Limit scope of disabling markdown support to just content

Use the pre insert filter on the rest endpoint to determine
if the content being saved is from Gutenberg and disable the
markdown support if a block is included.

Temporary includes the content_has_block() function from PR#2806

mkaz committed:
Move content_has_blocks to register.php to be compat with PR #2806

aduth commented on lib/register.php:
```diff
@@ -286,6 +286,18 @@ function gutenberg_post_has_blocks( $post_id ) {
+ * @param string $content Content to test.
+ * @return bool Whether the content contains blocks.
+ */
+function content_has_blocks( $content ) {
```
If we're introducing this function, I think we should refactor `gutenberg_post_has_blocks` to use it as well, rather than repeat the `strpos` implementation.

mkaz replied:
Good catch, updated `gutenberg_post_has_blocks` to use the new function

mkaz committed:
Refactor gutenberg_post_has_blocks to use content_has_blocks check

aaronjorbin commented on lib/compat.php:
```diff
@@ -132,10 +132,29 @@ function gutenberg_ensure_wp_api_request() {
+ * @param  array $post      Post object which contains content to check for block.
+ * @return array $post      Post object.
+ */
+function gutenberg_remove_wpcom_markdown_support( $post ) {
```
Shouldn't this live in the plugin that adds `wpcom-markdown` to posts?

mkaz replied:
@aaronjorbin it does makes sense to include in Jetpack. However, there are also other plugins, for example <a href="https://wordpress.org/plugins/jetpack-markdown/">JP Markdown</a> which include it.

aaronjorbin replied:
@mkaz What about the [other plugins](https://wordpress.org/plugins/tags/markdown/) that add markdown in a different way? Should the code for turning off markdown in each of them live in Gutenberg or should Gutenberg provide the tools (such as the `content_has_blocks` function and ideally some documentation) to assist them?

mkaz replied:
@aaronjorbin Gutenberg definitely should provide tools to assist, which this change does include one.

It's a bit of a catch-22, Gutenberg is still a plugin under development, trying to grow and get more people interested and gather feedback. If user wants to try it out and then their first experience is a broken, it's not a great way to gather feedback. So this fix disables a feature in one of the most popular plugins that makes for a poor Gutenberg experience.

As Gutenberg grows and gets closer to becoming the core editor, plugins will adopt the changes and include fixes themselves. However, as stated it is harder to grow if initial experience is rough.

Ideally, a plugin could provide a markdown block which would be awesome and any parsing conflict could also be avoided.

aaronjorbin replied:
I completely agree on the catch-22, but I think it also goes other ways. If Gutenberg is going to be in core, it can't be responsible for changes to plugins.  Either things need to be done in a fully backward compatible way, or education is needed for plugin maintainers to help them adjust. Playing favorites for popular plugins is going going to lead to unnecessary drama (what qualifies as a popular plugin?  Why was that plugin given preference over mine?), an increased and unnecessary maintenance burdon for core, and slower fixes for users.

Perhaps what's needed is some documentation and a documentation standard for things that can't be included in a core merge if that was to happen?

mkaz committed:
Move markdown fix to own plugin-compatibility file

Creating a new file as a means to better document compatibility
support added for specific plugins. This change also isolates code
that likely will not be included for a core merge.

mkaz commented:
@aaronjorbin - I added a file called `plugin-compat.php` which includes the plugin specific part of this change. I added documentation at the top which I hope captures our discussion. 

Let me know what you think, see commit <a href="https://github.com/WordPress/gutenberg/pull/2817/commits/520671b85253f09616cc6e83b449859b962490df">520671b</a>.

aaronjorbin commented:
@mkaz Looks good to me.  Seems to achieve the goal of improving compatibility to help with testing while also making a core merge easier.  Thanks for working on this and for making that change.
